### ISSUE TITLE
allow rule on @auth is not throwing FORBIDDEN error when rule is not satisfied on CREATE mutation

### ISSUE THREAD

**Issue by Karthikeyan-Vijayan** [neo4j/graphql#1095]
**Describe the bug**
I am trying to create a allow rule based on different user types to restrict some user types when we try to create Section nodes. But allow rule is not throwing FORBIDDEN error when rule is not satisfied.

**Type definitions**

```gql
type User {
    id: ID! @id
    name: String!
}

type Post {
    id: ID! @id
    name: String!
    info: String!
    author: User! @relationship(type: "CREATED", direction: IN)
    coAuthor: User @relationship(type: "CO_AUTHORED", direction: IN)
    moderators: [User!]! @relationship(type: "MODERATING", direction: IN)
}

extend type Post
    @auth(
        rules: [
            { operations: [READ], isAuthenticated: true }
            { operations: [CREATE], bind: { author: { id: "$jwt.userId" } } }
            {
                operations: [UPDATE]
                allow: { author: { id: "$jwt.userId" } }
                bind: { author: { id: "$jwt.userId" } }
            }
            { operations: [DELETE], allow: { author: { id: "$jwt.userId" } } }
        ]
    )

type Section {
    id: ID! @id
    name: String!
    post: Post @relationship(type: "HAVE_SECTION", direction: IN)
    createdBy: User! @relationship(type: "CREATED_SECTION", direction: IN)
}

extend type Section
    @auth(
        rules: [
            { operations: [READ], isAuthenticated: true }
            {
                operations: [CREATE]
                # AND:[{allow: {OR:[{post:{author: {id: "$jwt.userId"}}},
                # {post:{coAuthor: {id: "$jwt.userId"}}}
                # ]}},
                # {bind: { createdBy: {id: "$jwt.userId"} }
                # }],
                allow: {
                    OR: [
                        { post: { author: { id: "$jwt.userId" } } }
                        { post: { coAuthor: { id: "$jwt.userId" } } }
                    ]
                }
            }
            {
                operations: [UPDATE, DELETE]
                allow: {
                    OR: [{ post: { author: { id: "$jwt.userId" } } }, { createdBy: { id: "$jwt.userId" } }]
                }
            }
        ]
    )
```

**To Reproduce**

1. Run a server with the following code...
```
plugins: {
        auth: new Neo4jGraphQLAuthJWTPlugin({
            secret: token_secret
        })
    }
```
2. Execute the following Mutation...

```gql
mutation{
  createSections(input: [{
    name: "created_by_author",
    createdBy:
    {
      connect:
      {
        where: {node:{id: "f5242040-2e9d-473c-a581-90305b390735"}}
      }
    }
    post:
    {
      connect:
      {
        where: {node:{id: "550db793-bbcb-446f-aae0-db7aa757641d"}}
      }
    }
  }]){
    sections{
    name
    id
    createdBy {
      id
      name
    }
    post {
      id
      name
    } 
    }
  }
}
```

3. Then run the following Mutation...

```gql
mutation createSection{
  createSections(input: [{
    name: "created_by_moderator",
    createdBy:
    {
      connect:
      {
        where: {node:{id: "09a11adf-46de-452d-8626-61647256aa92"}}
      }
    }
    post:
    {
      connect:
      {
        where: {node:{id: "550db793-bbcb-446f-aae0-db7aa757641d"}}
      }
    }
  }]){
    sections{
    name
    id
    createdBy {
      id
      name
    }
    post{
      id
      name
    } 
    }
  }
}
```

**Expected behavior**
FORBIDDEN error should have thrown when executing the above mutation which creates Section nodes. Because allow rule will only permit users who are either author (mutation in 2) or coAuthor. Moderators are not allowed to create Sections (mutation in 3).

But FORBIDDEN error is not throwed when running mutation in 3. 
**System (please complete the following information):**

OS: Windows
Version: https://github.com/orgs/neo4j/teams/graphql@3.0.1
Node.js version: v17.3.1

**Comment by neo4j-team-graphql** [comment]
Many thanks for raising this bug report @Karthikeyan-Vijayan. :bug: We will now attempt to reproduce the bug based on the steps you have provided.

Please ensure that you've provided the necessary information for a minimal reproduction, including but not limited to:
* Type definitions
* Resolvers
* Query and/or Mutation (or multiple) needed to reproduce

If you have a support agreement with Neo4j, please link this GitHub issue to a new or existing Zendesk ticket.

Thanks again! :pray:

**Comment by angrykoala** [comment]
Hi @Karthikeyan-Vijayan 
I'm afraid that to faithfully reproduce this problem, I need to know more context regarding the users and posts in the database
I noticed that both mutations are the equal, maybe a copy-paste error?

**Comment by Karthikeyan-Vijayan** [comment]
Users are created using following mutation:
mutation($input: [UserCreateInput!]!) {
  createUsers(input: $input) {
    users {
      id
      name
    }
  }
}
{
  "input": [
    {
      "name": "test"
    },
    {
      "name": "test2"
    },
    {
      "name": "test3"
    }
  ]
}

Post will be created using following mutation where id of test is given to author, test2 is given to coAuthor and test3 is given to moderators :
mutation {
  createPosts(input: [{
    name: "test",
    info: "test",
    author:
    {
      connect:
      {
        where: {node:{id: "f5242040-2e9d-473c-a581-90305b390735"}}
      }
    }
    coAuthor:
    {
      connect:
      {
        where: {node:{id: "726b27ed-111d-4d31-ac31-13e090ca1eb1"}}
      }
    }
    moderators:
    {
      connect:
      {
        where: {node:{id: "09a11adf-46de-452d-8626-61647256aa92"}}
      }
    }
  }]){
    posts {
    name
    id
    author {
      id
      name
    }
    moderators {
      id
      name
    }
    coAuthor {
      id
      name
    }
    }
  }
}

And you are correct.  There is a copy paste mistake. mutation given in 2 on first comment will have id of author from post (in this case id of User node with name "test") for createdBy relationship. 

mutation in 3 on first comment will have id of moderator from post (in this case id of User node with name "test3") for createdBy relationship. This mutation needs to throw FORBIDDEN error because moderators are not allowed to create Section nodes.

I hope this helps.

**Comment by neo4j-team-graphql** [comment]
We've been able to confirm this bug using the steps to reproduce that you provided - many thanks @Karthikeyan-Vijayan! :pray: We will now prioritise the bug and address it appropriately.

**Comment by angrykoala** [comment]
Thanks @Karthikeyan-Vijayan !
If you could edit the original comment to fix mutation 2 that would be great

**Comment by Karthikeyan-Vijayan** [comment]
> We just want to reconfirm this against 3.0.2, as we did some bug fixing in the auth space.

I tried to reproduce this with version 3.0.3. The problem still exists. No FORBIDDEN error thrown.

**Comment by darrellwarde** [comment]
Sounds to be like this will be satisfied by a combination of #1871 and using a different argument - `bind`. 

You need to use `bind`, because before a `create` (which is when `allow` is evaluated), there is not yet a `post` to check against. So it needs to be checked after the `Section` has been created (which will be rolled back if authorization fails).

When implemented, #1871 will allow more complex logic in authorization rules. For instance:

```gql
extend type Section
    @auth(
        rules: [
            {
                operations: [CREATE],
                bind: {
                    AND: [
                        NOT: { post: { moderator: { id: "$jwt.userId" } } },
                        OR: [
                            { post: { author: { id: "$jwt.userId" } } }
                            { post: { coAuthor: { id: "$jwt.userId" } } }
                        ]
                    ]
                    
                }
            }
        ]
    )
```

The above rule would say "the moderator of the connected Post must NOT be the current user, AND the current user must be either the author or co-author of the connected Post". 

This is not yet possible, but is currently being designed. So I'm going to close this and add it to the list of issues which will be satisfied by #1871. Thanks!
